src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -1024,10 +1025,17 @@ void ReplicatedMergeTreeQueue::removePartProducingOpsInRange(
     [[maybe_unused]] bool called_from_alter_query_directly = covering_entry && covering_entry->replace_range_entry
         && covering_entry->replace_range_entry->columns_version < 0;
     [[maybe_unused]] bool called_for_broken_part = !covering_entry;
-    assert(currently_executing_drop_or_replace_range || called_from_alter_query_directly || called_for_broken_part);
+    assert(currently_executing_drop_or_replace_range || called_from_alter_query_directly || called_for_broken_part || !fetch_entry_znode.empty());
 
     for (Queue::iterator it = queue.begin(); it != queue.end();)
     {
+        /// Skipping currently processing entry
+        if (!fetch_entry_znode.empty() && (*it)->znode_name == fetch_entry_znode)
+        {
+            ++it;
+            continue;
+        }
```

Comment:
I'm not sure that it will work correctly with "actual part name" optimization...

Reply:
You mean that here `actual_new_part_name` should be take into account?
https://github.com/ClickHouse/ClickHouse/blob/8d1d2ecbfa0e06a212e52dd357a8fd2dc55233e5/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1037

Doesn't this means that `removePartProducingOpsInRange` already works not correctly when `actual_new_part_name` is not empty?

Reply:
> You mean that here actual_new_part_name should be take into account?

Not precisely, I mean that logic that "upgrades" currently executing entry to covering entry probably should be taken into account. See `addFuturePartIfNotCoveredByThem` and related methods. Seems like deadlock is possible, because `setActualPartName` may wait for older entries ("age" is determined by `znode_name`), but `removePartProducingOpsInRange` now may wait for all entries covered by `part_info`. And usually covered entries are older, but not when "actual part name" logic is involved.

> Doesn't this means that removePartProducingOpsInRange already works not correctly when actual_new_part_name is not empty?

No, because currently `removePartProducingOpsInRange` can be called by `DROP_RANGE` or `REPLACE_RANGE` entries only and `actual_new_part_name` can be set only for simple part-producing operation (`GET_PART`, `MERGE_PARTS`, `MUTATE_PART`)

Reply:
So I've looked at this again, and let me describe possible scenarios (that may lead to a possible deadlock):

- queue: [`1_0_1_0`, `1_1_2_0`, `1_0_2_1`]
  `T1` will execute fetch of `1_0_1_0`, and fetch `1_0_2_1` instead
  `T2` will execute fetch of `1_1_2_0`, no `1_0_2_1` in `future_parts` yet
  This should be safe because `T2` will not fetch anything because of https://github.com/ClickHouse/ClickHouse/blob/d4a7e1892caba0658d1fa8f9ac1244ed3c1d52ac/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1116-L1125

- queue: [`1_0_1_0`, `1_1_2_0`, `1_0_2_1`]
  `T1` will execute fetch of `1_0_1_0`, and fetch `1_0_2_1` instead
  `T2` will execute fetch of `1_1_2_0`, and also fetch `1_0_2_1` instead
  This should be safe because of https://github.com/ClickHouse/ClickHouse/blob/d4a7e1892caba0658d1fa8f9ac1244ed3c1d52ac/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1069-L1084

- queue: [`1_0_1_0`, `1_1_2_0`, `1_0_2_1`]
  `T1` will execute fetch of `1_0_1_0`, and fetch `1_0_2_1` instead
  `T2` will execute fetch of `1_0_2_1`
  This should be safe, because `actual_part_name` is also added into https://github.com/ClickHouse/ClickHouse/blob/d4a7e1892caba0658d1fa8f9ac1244ed3c1d52ac/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1492

So I don't see any possible deadlock, @tavplubix maybe you can suggest some example of it?

Reply:
@tavplubix friendly ping

Reply:
Sorry, I'm quite busy lately... But I remember about this PR and will take a look later

Reply:
@tavplubix just one more friendly ping (let me know if you don't need them, and you remember everything)

Reply:
@tavplubix thanks for the details and for finish the patch!

>Potential problem is that queue entries may be reordered during recovery of stale replica

But why can't they be reordered without recovery?

>and in rare cases covering entry may have less znode_name (you can check it by running test 993):

And if this is possible then it seems that this check is indeed redundant, no?

https://github.com/ClickHouse/ClickHouse/blob/0b4e643c2764543911b0b4c0c3745846af05742e/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1147

Reply:
> But why can't they be reordered without recovery?

It depends of what "order" is. There are at least 3 orders:
 - total order based on `znode_name`/sequential numbers of log/queue entries
 - partial (or total (?) within partition) order based on `MergeTreePartInfo` of parts produced by queue entries
 - "order" of entries execution (there's no order actually, it can be almost arbitrary)

This statement was about `znode_name` order. Entries for bigger (in terms of `MergeTreePartInfo`) parts usually get greater sequential numbers (and greater `znode_name`s). Just because when we create an entry for a covering part, entries for all source parts were already created in the log. And entries are pulled from log to queue one by one in the same order. 

But recovery of stale replica is tricky: it tries to copy queue of source replica (while this queue is concurrently changing) and additionally creates GET_PART entries for all parts that source replica had at the moment after copying queue entries. And an entry for bigger part can get less `znode_name` even if it had greater `znode_name` in source replica queue (just like in the log message above). Something similar may happen due to `removePartAndEnqueueFetch` without recovery.

> And if this is possible then it seems that this check is indeed redundant, no?

No, this check is needed to enforce _some_ order [here](https://github.com/ClickHouse/ClickHouse/blob/c64109a2dec3739b0ec1b0bdf82a4e47b4d75d28/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1526-L1532), comparing `znode_name` was the simplest way to avoid [deadlocks](https://s3.amazonaws.com/clickhouse-test-reports/37822/6ddfec47d2a71bc4d33c6a128f5dc254f972f95a/stateless_tests__release__actions_.html).

Reply:
>Something similar may happen due to removePartAndEnqueueFetch without recovery.

Oh, I see, thanks. Didn't know that queue can be filled not from the zookeeper log.

>No, this check is needed to enforce some order [here](https://github.com/ClickHouse/ClickHouse/blob/c64109a2dec3739b0ec1b0bdf82a4e47b4d75d28/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1526-L1532), comparing znode_name was the simplest way to avoid [deadlocks](https://s3.amazonaws.com/clickhouse-test-reports/37822/6ddfec47d2a71bc4d33c6a128f5dc254f972f95a/stateless_tests__release__actions_.html)

So let me describe how I understood it, and please correct me if I'm wrong.

Server tries to fetch `1_417_417_0` (`queue-0000004322`), but it had been "upgraded" to fetch `1_300_433_10` instead, but it still waiting for:
- `1_381_386_1` (`queue-0000004859`)
- `1_381_394_2` (`queue-0000004868`)
- `1_381_412_3` (`queue-0000004886`)
- `1_388_388_0` (`queue-0000004008`)
- `1_417_417_0` (`queue-0000004322`) -- will not try to wait it

After your patch it will wait only for `1_388_388_0`, but actually it will not start execution until it will be executed, and so deadlock is not possible after this?

Looks like this is happening because of the same "actual_part_name" optimization? And it is safe because this optimization can be done only locally, it is not stored in zookeeper, and so no issues if the queue entries will be reordered during replica clone. And it is also safe for the case with `removePartAndEnqueueFetch`, since it can create only parts with greater znode.

Maybe it worth putting a comment there (for znode comparison why it is safe, and what guarantees it has)?

Also maybe just for the sake of consistency it worth to make this check regardless `covered_entries_to_wait` was passed or not? (for the `shouldExecuteLogEntry`)

https://github.com/ClickHouse/ClickHouse/blob/0b4e643c2764543911b0b4c0c3745846af05742e/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1147-L1155